File: uniplex/session.py
"""
Uniplex Sessions - Managing ongoing authorization relationships.

Sessions allow:
- Reduced overhead for repeated authorizations
- Session-bound PoP for enhanced security
- Audit trails across related requests

Per spec: Session tokens use ISO 8601 timestamps, not epoch seconds.
"""

import json
import secrets
import uuid
from datetime import datetime, timedelta, timezone
from typing import Dict, Optional, Set

from pydantic import BaseModel, Field

from . import crypto
from .passport import Passport


class SessionToken(BaseModel):
    """
    Session token for ongoing authorization.
    
    Contains session metadata and can be used for session-bound PoP.
    """
    
    session_id: str = Field(default_factory=lambda: f"ses_{uuid.uuid4().hex[:16]}")
    nonce: str = Field(default_factory=lambda: secrets.token_urlsafe(16))
    
    # Passport binding
    passport_id: str
    agent_id: str
    
    # Target binding
    target: Optional[str] = None
    
    # Timing (ISO 8601 per spec)
    iat: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    )
    exp: str  # Required - when session expires
    
    # Permissions granted for this session
    permissions: list[str] = Field(default_factory=list)
    
    # Metadata
    metadata: Dict = Field(default_factory=dict)
    
    @classmethod
    def create(
        cls,
        passport: Passport,
        target: Optional[str] = None,
        duration: timedelta = timedelta(hours=1),
        permissions: Optional[list[str]] = None,
    ) -> "SessionToken":
        """
        Create a new session token.
        
        Args:
            passport: The passport to bind the session to
            target: Optional target binding
            duration: Session duration (default 1 hour)
            permissions: Permissions for this session (default: passport permissions)
        
        Returns:
            New SessionToken
        """
        now = datetime.now(timezone.utc)
        exp = now + duration
        
        # Default to passport permissions
        if permissions is None:
            permissions = [p.action for p in passport.permissions]
        
        return cls(
            passport_id=passport.passport_id,
            agent_id=passport.identity.agent_id,
            target=target,
            exp=exp.strftime("%Y-%m-%dT%H:%M:%SZ"),
            permissions=permissions,
        )
    
    def is_expired(self) -> bool:
        """Check if the session has expired."""
        try:
            exp = datetime.fromisoformat(self.exp.replace("Z", "+00:00"))
            return datetime.now(timezone.utc) > exp
        except Exception:
            return True
    
    def is_valid_for(self, passport_id: str, action: Optional[str] = None) -> bool:
        """
        Check if session is valid for a given passport and action.
        
        Args:
            passport_id: Passport ID to check
            action: Optional action to check permission for
        
        Returns:
            True if session is valid
        """
        if self.is_expired():
            return False
        
        if self.passport_id != passport_id:
            return False
        
        if action and not self._has_permission(action):
            return False
        
        return True
    
    def _has_permission(self, action: str) -> bool:
        """Check if session grants permission for action."""
        for perm in self.permissions:
            if perm == "*":
                return True
            if perm == action:
                return True
            if perm.endswith(":*"):
                prefix = perm[:-1]
                if action.startswith(prefix):
                    return True
        return False
    
    def to_dict(self) -> dict:
        """Export as dictionary."""
        return self.model_dump(exclude_none=True)
    
    def to_json(self) -> str:
        """Export as JSON string."""
        return json.dumps(self.to_dict())
    
    @classmethod
    def from_dict(cls, data: dict) -> "SessionToken":
        """Create from dictionary."""
        return cls.model_validate(data)
    
    @classmethod
    def from_json(cls, json_str: str) -> "SessionToken":
        """Create from JSON string."""
        return cls.from_dict(json.loads(json_str))


class SessionManager:
    """
    Manager for authorization sessions.
    
    Provides session creation, validation, and cleanup.
    """
    
    def __init__(
        self,
        default_duration: timedelta = timedelta(hours=1),
        max_sessions_per_agent: int = 10,
    ):
        """
        Create a session manager.
        
        Args:
            default_duration: Default session duration
            max_sessions_per_agent: Maximum concurrent sessions per agent
        """
        self.default_duration = default_duration
        self.max_sessions_per_agent = max_sessions_per_agent
        
        self._sessions: Dict[str, SessionToken] = {}  # session_id -> token
        self._agent_sessions: Dict[str, Set[str]] = {}  # agent_id -> set of session_ids
        self._revoked: Set[str] = set()  # Revoked session IDs
    
    def create_session(
        self,
        passport: Passport,
        target: Optional[str] = None,
        duration: Optional[timedelta] = None,
        permissions: Optional[list[str]] = None,
    ) -> SessionToken:
        """
        Create a new session for a passport.
        
        Args:
            passport: The passport to create session for
            target: Optional target binding
            duration: Session duration (uses default if not provided)
            permissions: Session permissions (uses passport permissions if not provided)
        
        Returns:
            New SessionToken
        """
        # Check max sessions
        agent_id = passport.identity.agent_id
        if agent_id in self._agent_sessions:
            if len(self._agent_sessions[agent_id]) >= self.max_sessions_per_agent:
                # Clean up expired sessions first
                self._cleanup_agent_sessions(agent_id)
                # Check again
                if len(self._agent_sessions.get(agent_id, set())) >= self.max_sessions_per_agent:
                    raise ValueError(f"Maximum sessions reached for agent {agent_id}")
        
        # Create session
        token = SessionToken.create(
            passport=passport,
            target=target,
            duration=duration or self.default_duration,
            permissions=permissions,
        )
        
        # Store
        self._sessions[token.session_id] = token
        if agent_id not in self._agent_sessions:
            self._agent_sessions[agent_id] = set()
        self._agent_sessions[agent_id].add(token.session_id)
        
        return token
    
    def get_session(self, session_id: str) -> Optional[SessionToken]:
        """Get a session by ID."""
        if session_id in self._revoked:
            return None
        return self._sessions.get(session_id)
    
    def validate_session(
        self,
        session_id: str,
        passport_id: str,
        action: Optional[str] = None,
    ) -> tuple[bool, Optional[str]]:
        """
        Validate a session.
        
        Args:
            session_id: Session ID to validate
            passport_id: Expected passport ID
            action: Optional action to check permission for
        
        Returns:
            Tuple of (is_valid, error_message)
        """
        if session_id in self._revoked:
            return False, "SESSION_REVOKED"
        
        session = self._sessions.get(session_id)
        if not session:
            return False, "SESSION_NOT_FOUND"
        
        if session.is_expired():
            return False, "SESSION_EXPIRED"
        
        if session.passport_id != passport_id:
            return False, "SESSION_PASSPORT_MISMATCH"
        
        if action and not session._has_permission(action):
            return False, "SESSION_PERMISSION_DENIED"
        
        return True, None
    
    def revoke_session(self, session_id: str) -> bool:
        """
        Revoke a session.
        
        Args:
            session_id: Session to revoke
        
        Returns:
            True if session was found and revoked
        """
        if session_id in self._sessions:
            self._revoked.add(session_id)
            session = self._sessions[session_id]
            if session.agent_id in self._agent_sessions:
                self._agent_sessions[session.agent_id].discard(session_id)
            del self._sessions[session_id]
            return True
        return False
    
    def revoke_all_for_agent(self, agent_id: str) -> int:
        """
        Revoke all sessions for an agent.
        
        Returns:
            Number of sessions revoked
        """
        session_ids = self._agent_sessions.get(agent_id, set()).copy()
        count = 0
        for session_id in session_ids:
            if self.revoke_session(session_id):
                count += 1
        return count
    
    def _cleanup_agent_sessions(self, agent_id: str) -> None:
        """Remove expired sessions for an agent."""
        if agent_id not in self._agent_sessions:
            return
        
        expired = []
        for session_id in self._agent_sessions[agent_id]:
            session = self._sessions.get(session_id)
            if not session or session.is_expired():
                expired.append(session_id)
        
        for session_id in expired:
            self._agent_sessions[agent_id].discard(session_id)
            self._sessions.pop(session_id, None)
    
    def cleanup_expired(self) -> int:
        """
        Remove all expired sessions.
        
        Returns:
            Number of sessions removed
        """
        expired = []
        for session_id, session in self._sessions.items():
            if session.is_expired():
                expired.append(session_id)
        
        for session_id in expired:
            session = self._sessions.pop(session_id, None)
            if session and session.agent_id in self._agent_sessions:
                self._agent_sessions[session.agent_id].discard(session_id)
        
        return len(expired)
    
    def get_agent_sessions(self, agent_id: str) -> list[SessionToken]:
        """Get all active sessions for an agent."""
        self._cleanup_agent_sessions(agent_id)
        session_ids = self._agent_sessions.get(agent_id, set())
        return [self._sessions[sid] for sid in session_ids if sid in self._sessions]


# Global session manager
_default_manager: Optional[SessionManager] = None


def get_session_manager() -> SessionManager:
    """Get the default session manager."""
    global _default_manager
    if _default_manager is None:
        _default_manager = SessionManager()
    return _default_manager


def set_session_manager(manager: SessionManager) -> None:
    """Set the default session manager."""
    global _default_manager
    _default_manager = manager

File: js/src/session.ts
/**
 * Uniplex Sessions - Managing ongoing authorization relationships.
 *
 * Sessions allow:
 * - Reduced overhead for repeated authorizations
 * - Session-bound PoP for enhanced security
 * - Audit trails across related requests
 */

import { Passport } from './passport.js';

// ============================================================================
// Types
// ============================================================================

export interface SessionTokenData {
  session_id: string;
  nonce: string;
  passport_id: string;
  agent_id: string;
  target?: string;
  iat: string;
  exp: string;
  permissions: string[];
  metadata?: Record<string, unknown>;
}

// ============================================================================
// Helper Functions
// ============================================================================

function generateSessionId(): string {
  const chars = 'abcdef0123456789';
  let id = 'ses_';
  for (let i = 0; i < 16; i++) {
    id += chars[Math.floor(Math.random() * chars.length)];
  }
  return id;
}

function generateNonce(): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
  let nonce = '';
  for (let i = 0; i < 22; i++) {
    nonce += chars[Math.floor(Math.random() * chars.length)];
  }
  return nonce;
}

// ============================================================================
// SessionToken Class
// ============================================================================

export class SessionToken {
  readonly sessionId: string;
  readonly nonce: string;
  readonly passportId: string;
  readonly agentId: string;
  readonly target?: string;
  readonly iat: string;
  readonly exp: string;
  readonly permissions: string[];
  readonly metadata?: Record<string, unknown>;

  constructor(data: SessionTokenData) {
    this.sessionId = data.session_id;
    this.nonce = data.nonce;
    this.passportId = data.passport_id;
    this.agentId = data.agent_id;
    this.target = data.target;
    this.iat = data.iat;
    this.exp = data.exp;
    this.permissions = data.permissions;
    this.metadata = data.metadata;
  }

  /**
   * Create a new session token.
   */
  static create(
    passport: Passport,
    options: {
      target?: string;
      durationMs?: number;
      permissions?: string[];
    } = {}
  ): SessionToken {
    const { target, durationMs = 60 * 60 * 1000, permissions } = options;

    const now = new Date();
    const exp = new Date(now.getTime() + durationMs);

    // Default to passport permissions
    const perms = permissions ?? passport.permissions.map((p) => p.action);

    return new SessionToken({
      session_id: generateSessionId(),
      nonce: generateNonce(),
      passport_id: passport.passportId,
      agent_id: passport.identity.agent_id,
      target,
      iat: now.toISOString().replace(/\.\d{3}Z$/, 'Z'),
      exp: exp.toISOString().replace(/\.\d{3}Z$/, 'Z'),
      permissions: perms,
    });
  }

  /**
   * Check if the session has expired.
   */
  isExpired(): boolean {
    const expDate = new Date(this.exp);
    return new Date() > expDate;
  }

  /**
   * Check if session is valid for a given passport and action.
   */
  isValidFor(passportId: string, action?: string): boolean {
    if (this.isExpired()) {
      return false;
    }

    if (this.passportId !== passportId) {
      return false;
    }

    if (action && !this.hasPermission(action)) {
      return false;
    }

    return true;
  }

  /**
   * Check if session grants permission for action.
   */
  hasPermission(action: string): boolean {
    for (const perm of this.permissions) {
      if (perm === '*') {
        return true;
      }
      if (perm === action) {
        return true;
      }
      if (perm.endsWith(':*')) {
        const prefix = perm.slice(0, -1);
        if (action.startsWith(prefix)) {
          return true;
        }
      }
    }
    return false;
  }

  /**
   * Convert to plain object.
   */
  toDict(): SessionTokenData {
    return {
      session_id: this.sessionId,
      nonce: this.nonce,
      passport_id: this.passportId,
      agent_id: this.agentId,
      target: this.target,
      iat: this.iat,
      exp: this.exp,
      permissions: this.permissions,
      metadata: this.metadata,
    };
  }

  /**
   * Convert to JSON string.
   */
  toJson(): string {
    return JSON.stringify(this.toDict(), null, 2);
  }

  /**
   * Create from plain object.
   */
  static fromDict(data: SessionTokenData): SessionToken {
    return new SessionToken(data);
  }

  /**
   * Create from JSON string.
   */
  static fromJson(json: string): SessionToken {
    return SessionToken.fromDict(JSON.parse(json));
  }
}

// ============================================================================
// SessionManager Class
// ============================================================================

export class SessionManager {
  readonly defaultDurationMs: number;
  readonly maxSessionsPerAgent: number;

  private sessions: Map<string, SessionToken> = new Map();
  private agentSessions: Map<string, Set<string>> = new Map();
  private revoked: Set<string> = new Set();

  constructor(
    options: {
      defaultDurationMs?: number;
      maxSessionsPerAgent?: number;
    } = {}
  ) {
    this.defaultDurationMs = options.defaultDurationMs ?? 60 * 60 * 1000;
    this.maxSessionsPerAgent = options.maxSessionsPerAgent ?? 10;
  }

  /**
   * Create a new session for a passport.
   */
  createSession(
    passport: Passport,
    options: {
      target?: string;
      durationMs?: number;
      permissions?: string[];
    } = {}
  ): SessionToken {
    const agentId = passport.identity.agent_id;

    // Check max sessions
    const existingSessions = this.agentSessions.get(agentId);
    if (existingSessions && existingSessions.size >= this.maxSessionsPerAgent) {
      this.cleanupAgentSessions(agentId);
      const updated = this.agentSessions.get(agentId);
      if (updated && updated.size >= this.maxSessionsPerAgent) {
        throw new Error(`Maximum sessions reached for agent ${agentId}`);
      }
    }

    // Create session
    const session = SessionToken.create(passport, {
      target: options.target,
      durationMs: options.durationMs ?? this.defaultDurationMs,
      permissions: options.permissions,
    });

    // Store
    this.sessions.set(session.sessionId, session);
    if (!this.agentSessions.has(agentId)) {
      this.agentSessions.set(agentId, new Set());
    }
    this.agentSessions.get(agentId)!.add(session.sessionId);

    return session;
  }

  /**
   * Get a session by ID.
   */
  getSession(sessionId: string): SessionToken | undefined {
    if (this.revoked.has(sessionId)) {
      return undefined;
    }
    return this.sessions.get(sessionId);
  }

  /**
   * Validate a session.
   */
  validateSession(
    sessionId: string,
    passportId: string,
    action?: string
  ): { valid: boolean; error?: string } {
    if (this.revoked.has(sessionId)) {
      return { valid: false, error: 'SESSION_REVOKED' };
    }

    const session = this.sessions.get(sessionId);
    if (!session) {
      return { valid: false, error: 'SESSION_NOT_FOUND' };
    }

    if (session.isExpired()) {
      return { valid: false, error: 'SESSION_EXPIRED' };
    }

    if (session.passportId !== passportId) {
      return { valid: false, error: 'SESSION_PASSPORT_MISMATCH' };
    }

    if (action && !session.hasPermission(action)) {
      return { valid: false, error: 'SESSION_PERMISSION_DENIED' };
    }

    return { valid: true };
  }

  /**
   * Revoke a session.
   */
  revokeSession(sessionId: string): boolean {
    const session = this.sessions.get(sessionId);
    if (session) {
      this.revoked.add(sessionId);
      this.sessions.delete(sessionId);
      const agentSessions = this.agentSessions.get(session.agentId);
      if (agentSessions) {
        agentSessions.delete(sessionId);
      }
      return true;
    }
    return false;
  }

  /**
   * Revoke all sessions for an agent.
   */
  revokeAllForAgent(agentId: string): number {
    const sessionIds = this.agentSessions.get(agentId);
    if (!sessionIds) {
      return 0;
    }

    let count = 0;
    for (const sessionId of Array.from(sessionIds)) {
      if (this.revokeSession(sessionId)) {
        count++;
      }
    }
    return count;
  }

  /**
   * Clean up expired sessions for an agent.
   */
  private cleanupAgentSessions(agentId: string): void {
    const sessionIds = this.agentSessions.get(agentId);
    if (!sessionIds) {
      return;
    }

    const expired: string[] = [];
    for (const sessionId of sessionIds) {
      const session = this.sessions.get(sessionId);
      if (!session || session.isExpired()) {
        expired.push(sessionId);
      }
    }

    for (const sessionId of expired) {
      sessionIds.delete(sessionId);
      this.sessions.delete(sessionId);
    }
  }

  /**
   * Clean up all expired sessions.
   */
  cleanupExpired(): number {
    const expired: string[] = [];
    for (const [sessionId, session] of this.sessions) {
      if (session.isExpired()) {
        expired.push(sessionId);
      }
    }

    for (const sessionId of expired) {
      const session = this.sessions.get(sessionId);
      this.sessions.delete(sessionId);
      if (session) {
        const agentSessions = this.agentSessions.get(session.agentId);
        if (agentSessions) {
          agentSessions.delete(sessionId);
        }
      }
    }

    return expired.length;
  }

  /**
   * Get all active sessions for an agent.
   */
  getAgentSessions(agentId: string): SessionToken[] {
    this.cleanupAgentSessions(agentId);
    const sessionIds = this.agentSessions.get(agentId);
    if (!sessionIds) {
      return [];
    }

    const sessions: SessionToken[] = [];
    for (const sessionId of sessionIds) {
      const session = this.sessions.get(sessionId);
      if (session) {
        sessions.push(session);
      }
    }
    return sessions;
  }
}

// ============================================================================
// Default Manager
// ============================================================================

let defaultManager: SessionManager | undefined;

export function getSessionManager(): SessionManager {
  if (!defaultManager) {
    defaultManager = new SessionManager();
  }
  return defaultManager;
}

export function setSessionManager(manager: SessionManager): void {
  defaultManager = manager;
}
